src/hotspot/share/gc/shenandoah/shenandoahClosures.inline.hpp

```diff
@@ -129,46 +129,28 @@ void ShenandoahEvacuateUpdateMetadataClosure<MO>::do_oop_work(T* p) {
       shenandoah_assert_marked(p, obj);
       oop resolved = ShenandoahBarrierSet::resolve_forwarded_not_null(obj);
       if (resolved == obj) {
-        resolved = _heap->evacuate_object(obj, _thread);
+        resolved = _heap->evacuate_object(obj, t);
+      }
+      if (atomic) {
+        ShenandoahHeap::atomic_update_oop(resolved, p, o);
+      } else {
+        RawAccess<IS_NOT_NULL | MO_UNORDERED>::oop_store(p, resolved);
       }
-      RawAccess<IS_NOT_NULL | MO>::oop_store(p, resolved);
     }
   }
 }
-template <DecoratorSet MO>
-void ShenandoahEvacuateUpdateMetadataClosure<MO>::do_oop(oop* p) {
-  do_oop_work(p);
+template <bool atomic>
+void ShenandoahEvacuateUpdateRootClosureBase<atomic>::do_oop(oop* p) {
+  do_oop_work(p, Thread::current());
```

Comment:
Not fond of calling `Thread::current()` on this hotpath. From the code, it looks like we only need it when actually calling `_heap->evacuate_object()`, can we move it there? 

Related: is `ShenandoahContextEvacuateUpdateRootsClosure::_thread` used now? Can we use it on some paths too?

Reply:
My mistake, fixed.